pulsar-client-cpp/lib/MultiTopicsConsumerImpl.cc

```diff
@@ -0,0 +1,621 @@
+        }
+        multiTopicsConsumerCreatedPromise_.setValue(shared_from_this());
+        return;
+    }
+}
+
+// subscribe for passed in topic
+Future<Result, Consumer> MultiTopicsConsumerImpl::subscribeOneTopicAsync(const std::string& topic) {
+    TopicNamePtr topicName;
+    ConsumerSubResultPromisePtr topicPromise = boost::make_shared<Promise<Result, Consumer>>();
+    if (!(topicName = TopicName::get(topic))) {
+        LOG_ERROR("TopicName invalid: " << topic);
+        topicPromise->setFailed(ResultInvalidTopicName);
+        return topicPromise->getFuture();
+    }
+
+    if (state_ == Closed || state_ == Closing) {
+        LOG_ERROR("MultiTopicsConsumer already closed when subscribe.");
+        topicPromise->setFailed(ResultAlreadyClosed);
+        return topicPromise->getFuture();
+    }
+
+    // subscribe for each partition, when all partitions completed, complete promise
+    lookupServicePtr_->getPartitionMetadataAsync(topicName).addListener(
+        boost::bind(&MultiTopicsConsumerImpl::subscribeTopicPartitions, shared_from_this(), _1, _2, topicName,
+                    subscriptionName_, conf_, topicPromise));
+    return topicPromise->getFuture();
+}
+
+void MultiTopicsConsumerImpl::subscribeTopicPartitions(const Result result,
+                                                       const LookupDataResultPtr partitionMetadata,
+                                                       TopicNamePtr topicName,
+                                                       const std::string& consumerName,
+                                                       ConsumerConfiguration conf,
+                                                       ConsumerSubResultPromisePtr topicSubResultPromise) {
+    if (result != ResultOk) {
+        LOG_ERROR("Error Checking/Getting Partition Metadata while MultiTopics Subscribing- " << result)
+        topicSubResultPromise->setFailed(result);
+        return;
+    }
+
+    boost::shared_ptr<ConsumerImpl> consumer;
+    ConsumerConfiguration config;
+    ExecutorServicePtr internalListenerExecutor = client_->getPartitionListenerExecutorProvider()->get();
+
+    // all the consumers should have same name.
+    config.setConsumerName(conf_.getConsumerName());
+    config.setConsumerType(conf_.getConsumerType());
+    config.setBrokerConsumerStatsCacheTimeInMs(conf_.getBrokerConsumerStatsCacheTimeInMs());
+    config.setMessageListener(
+        boost::bind(&MultiTopicsConsumerImpl::messageReceived, shared_from_this(), _1, _2));
+    config.setReceiverQueueSize(conf_.getReceiverQueueSize());
+
+    int partitionsNumber = partitionMetadata->getPartitions() >= 1 ? partitionMetadata->getPartitions() : 1;
+
+    Lock lock(mutex_);
+    topicsPartitions_.insert(std::make_pair(topicName->toString(), partitionsNumber));
+    lock.unlock();
+    allTopicPartitionsNumber_->fetch_add(partitionsNumber);
+
+    boost::shared_ptr<std::atomic<int>> partitionsNeedCreate =
```

Comment:
Why are you creating a consumer per partition, rather than just creating a PartitionedConsumerImpl for the whole topic?

Reply:
Thanks, using the same pattern as java client.

Reply:
Then the question follows, why is the java client doing this?